code/assignTime.m

```diff
@@ -0,0 +1,279 @@
+        elseif iChunk == length(allFlaggedIndices)
+            chunkIndices{counter} = allFlaggedIndices(currentStartIndex) + 1:allFlaggedIndices(currentStartIndex + 1);
+            chunkIndices{counter + 1} = allFlaggedIndices(currentStartIndex + 1) + 1:numPackets;
+        else
+            chunkIndices{counter} = allFlaggedIndices(currentStartIndex) + 1:allFlaggedIndices(currentStartIndex + 1);
+            currentStartIndex = currentStartIndex + 1;
+            counter = counter + 1;
+        end
+     end
+else
+    % No identified missing packets, all packets in one chunk
+    chunkIndices{1} = 1:numPackets;
+end
+
+%%
+% Loop through each chunk to determine offset to apply (as determined by
+% average difference between packetGenTime and expectedElapsed)
+disp('Determining start time of each chunk')
+
+% PacketGenTime in ms; convert difference to 1e-4 seconds, units of
+% systemTick and expectedElapsed
+diff_PacketGenTime = [1; diff(dataTable.PacketGenTime) * 1e1];
+
+numChunks = length(chunkIndices);
+chunksToExclude = [];
+meanError = NaN(1,numChunks);
+for iChunk = 1:numChunks
+    currentTimestampIndices = chunkIndices{iChunk};
+    
+    % Chunks must have at least 2 packets in order to have a valid
+    % diff_systemTick -- thus if chunk only one packet, it must be excluded
+    if length(currentTimestampIndices) == 1
+        chunksToExclude = [chunksToExclude iChunk];
+    end
+    % Always exclude the first packet of the chunk, because don't have an
+    % accurate diff_systemTick value for this first packet
+    currentTimestampIndices = currentTimestampIndices(2:end);
+    
+    % Differences between adjacent PacketGenTimes (in units of 1e-4
+    % seconds)
+    error = expectedElapsed(currentTimestampIndices) - diff_PacketGenTime(currentTimestampIndices);
+    meanError(iChunk) = median(error);
+end
+%%
+% Create corrected timing for each chunk
+counter = 1;
+for iChunk = 1:numChunks
+    if ~ismember(iChunk,chunksToExclude)
+        alignTime = dataTable.PacketGenTime(chunkIndices{iChunk}(1));
+        % alignTime in ms; meanError in units of systemTick
+        correctedAlignTime(counter) = alignTime + meanError(iChunk)*1e-1;
+        % TO DO: add or subtract meanError above??
+        counter = counter + 1;
+    end
+end
+
+%%
+% Indices in chunkIndices correspond to packets in dataTable.
+% CorrectedAlignTime corresponds to first packet for each chunk in
+% chunkIndices. Remove chunks identified above
+chunkIndices(chunksToExclude) = [];
+correctedAlignTime(find(isnan(correctedAlignTime))) = [];
+
+% correctedAlignTime is shifted slightly to keep times exactly aligned to
+% sampling rate; use maxFs for alignment points. In other words, take first
+% correctedAlignTime, and place all other correctedAlignTimes at multiples of
+% (1/Fs)
+deltaTime = 1/maxFs * 1000; % in milliseconds
+
+disp('Shifting chunks to align with sampling rate')
+multiples = floor(((correctedAlignTime - correctedAlignTime(1))/deltaTime) + 0.5);
```

Comment:
This is a lot faster than the previous code version! However, this solution may have some errors in a handful of locations. I'd like to know your opinion and present my solution below.

I will be using this data set on Box as an example: RCS_long_data_examples/Session1568667718731/DeviceNPC700239H/RawDataTD.json

When running your code and comparing the output of "correctedAlignTime" and "correctAlignTime_shifted" an interesting decision was made by the code in question at index 5. correctAlignTime(5) = ...0454394 and correctAlignTime_shifted(5) = ...0454396. In the original code where you search for the minimum, the time that was assigned to this index was ...0454392. From the values listed here, the time difference of the original method and the new method are both 2 away from the correctedAlignTime value. Which value would you prefer to assign? The time point before the correctAlignTime value, like in the original method, or the new method?

If you'd prefer the original method here is the code suggestion. I modified the previous version of assignTime.m

> allPossibleTimes = correctedAlignTime(1):deltaTime:correctedAlignTime(end) + deltaTime;
correctedAlignTime_shifted = NaN(size(correctedAlignTime));
correctedAlignTime_shifted(1) = correctedAlignTime(1);
for iTime = 2:length(correctedAlignTime)
    currentTime = correctedAlignTime(iTime);
    indSearch(1) = floor((currentTime-correctedAlignTime(1))/deltaTime)+1;
    indSearch(2) = ceil((currentTime-correctedAlignTime(1))/deltaTime)+1;
    if abs(currentTime-allPossibleTimes(indSearch(1))) < abs(currentTime-allPossibleTimes(indSearch(2)))
        indexOfPossibleTimes = indSearch(1);
    elseif abs(currentTime-allPossibleTimes(indSearch(1))) > abs(currentTime-allPossibleTimes(indSearch(2)))
        indexOfPossibleTimes = indSearch(2);
    else 
        [~,minInd] = min(indSearch-currentTime);
        indexOfPossibleTimes = indSearch(minInd);
    end
    correctedAlignTime_shifted(iTime) = allPossibleTimes(indexOfPossibleTimes);
end 

I timed this version and compared it to your solution.
Your solution: 0.000444 seconds
My solution: 0.024021 seconds
Your solution is clearly faster!

Reply:
I should also add that my solution reproduces the output of the old method of using the minimum function

Reply:
@kenlouie This is a good question -- in the newer implementation when a value is equidistant from both neighboring candidates, it will always go with the larger option (because of the 0.5 we're adding before the floor function). I think this is fine, since it will be consistent. I'd opt for the faster implementation. What do you think?

A side note - I think this method should work regardless if we stick with assuming Fs is faithful to 250, 500, or 1000Hz or if we decide to shift these values (e.g. 1000 --> 999.98Hz). Do you agree with this as well?

Reply:
I would opt for the faster implementation as well since it is a consistent choice.

Yes, I agree this method will work even with different integer and non-integer sampling frequencies. For non-integer sampling frequencies, the addition of 0.5 should result with no duplicate time stamps.